Refactored sig api following nist-branch (issue380)

christianpaquin opened the pull request:
Refactored sig api following nist-branch, to harmonize with new KEM api; keeping however the sign/verify api since it's easier to use in applications (goal of the master branch). Fixes issue #380.

christianpaquin commented:
Worked done, ready to merge. @dstebila you can take a look.

dstebila commented on src/sig/picnic/sig_picnic.h:
```diff
@@ -0,0 +1,90 @@
+
+OQS_SIG *OQS_SIG_picnic_L1_FS_new();
+
+extern OQS_STATUS OQS_SIG_picnic_L1_FS_keypair(uint8_t *public_key, uint8_t *secret_key);
```
I don't think we need the `extern` keyword here or anywhere else in this file.

christianpaquin replied:
This came from the nist-branch; should we remove it on both branches?

dstebila replied:
There was a recent PR on nist-branch #387 that removed a bunch of superfluous extern's, but perhaps they missed some.  Please remove it here, and I'll ask Ben to do the same on nist-branch.

truth-quark replied:
I think a few algorithms didn't get cleaned of externs as there was active work on them. Chances are I forgot to add it in the PR while avoiding potential merge conflicts/jumbled history.

dstebila commented on src/sig/qtesla/sig_qtesla.h:
```diff
@@ -0,0 +1,48 @@
+
+OQS_SIG *OQS_SIG_qTESLA_I_new();
+
+extern OQS_STATUS OQS_SIG_qTESLA_I_keypair(uint8_t *public_key, uint8_t *secret_key);
```
I don't think we need the `extern` keyword here or anywhere else in this file.

christianpaquin replied:
This came from the nist-branch; should we remove it on both branches?

dstebila replied:
There was a recent PR on nist-branch #387 that removed a bunch of superfluous extern's, but perhaps they missed some. Please remove it here, and I'll ask Ben to do the same on nist-branch.

dstebila commented on src/sig/sig.c:
```diff
@@ -1,79 +1,120 @@
-OQS_API OQS_STATUS OQS_SIG_keygen(const OQS_SIG *s, uint8_t *priv, uint8_t *pub) {
-	if (s == NULL) {
+OQS_STATUS OQS_SIG_keypair(const OQS_SIG *sig, uint8_t *public_key, uint8_t *secret_key) {
+	if (sig == NULL || sig->keypair(public_key, secret_key) != 0) {
```
Why are you comparing the output of `sig->keypair` with 0 as opposed to `OQS_SUCCESS`?

christianpaquin replied:
Copy-pasted from nist-branch; fixed here; should fix it there too...

christianpaquin replied:
Also created PR #393 on nist-branch to fix this

dstebila commented on src/sig/sig.c:
```diff
@@ -1,79 +1,120 @@
-OQS_API OQS_STATUS OQS_SIG_sign(const OQS_SIG *s, const uint8_t *priv, const uint8_t *msg, const size_t msg_len, uint8_t *sig, size_t *sig_len) {
-	if (s == NULL) {
+OQS_STATUS OQS_SIG_sign(const OQS_SIG *sig, uint8_t *signature, size_t *signature_len, const uint8_t *message, size_t message_len, const uint8_t *secret_key) {
+	if (sig == NULL || sig->sign(signature, signature_len, message, message_len, secret_key) != 0) {
```
Why are you comparing the output of `sig->sign` with 0 as opposed to `OQS_SUCCESS`?

christianpaquin replied:
Copy-pasted from nist-branch; fixed here; should fix it there too...

christianpaquin replied:
Also created PR #393 on nist-branch to fix this

dstebila commented on src/sig/sig.c:
```diff
@@ -1,79 +1,120 @@
-OQS_API OQS_STATUS OQS_SIG_verify(const OQS_SIG *s, const uint8_t *pub, const uint8_t *msg, const size_t msg_len, const uint8_t *sig, const size_t sig_len) {
-	if (s == NULL) {
+OQS_STATUS OQS_SIG_verify(const OQS_SIG *sig, uint8_t *message, size_t message_len, const uint8_t *signature, size_t signature_len, const uint8_t *public_key) {
+	if (sig == NULL || sig->verify(message, message_len, signature, signature_len, public_key) != 0) {
```
Why are you comparing the output of `sig->verify` with 0 as opposed to `OQS_SUCCESS`?

christianpaquin replied:
Copy-pasted from nist-branch; fixed here; should fix it there too...

christianpaquin replied:
Also created PR #393 on nist-branch to fix this

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
 	size_t length_secret_key;
-	/** The (maximum) length, in bytes, of a signature for this signature scheme. */
-	size_t max_length_signature;
+	/** The (maximum) length, in bytes, of signatures for this signature scheme. */
```
Let's take the (maximum) out of parentheses.

christianpaquin replied:
I copied that from kem.h; we should keep or change both...

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
-	/** The (maximum) length, in bytes, of a signature for this signature scheme. */
-	size_t max_length_signature;
+	/** The (maximum) length, in bytes, of signatures for this signature scheme. */
+	size_t length_signature;
```
I wonder if we should call this `length_signature_max` instead of `length_signature`, since it's not necessarily the length of the signature but the maximum length of the signature.

christianpaquin replied:
given that maximum was implied in the comment, and that applied to the keys as well, I didn't want to single the signature field with a special "max" marker in its name

dstebila replied:
Okay, makes sense.

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
-	 * @param sig              The generated signature. Must be allocated by the caller, or NULL to learn how much space is needed, as returned in sig_len.
-	 * @param sig_len          In: length of sig, out: length of the generated signature.
-	 * @return                 OQS_SUCCESS on success, or OQS_ERROR on failure.
+	 * Caller is responsible for allocating sufficient memory for `signed_message`,
```
Change documentation to remove `signed_message`.

Also, should we have a comment explicitly explaining that in this branch `signature` contains just the digest, whereas in nist-branch `signature` contains both the digest and the message?

christianpaquin replied:
Changed. We need to explain it somewhere, but I'm not sure where. For folks using the master branch for integration, it might not make sense to talk about a "nist-branch"; and the signature/verify paradigm used by the master branch is the classic API for signatures.

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
-	 * Pointer to a function for signature verification.
+	 * Signature verification algorithm.
+	 *
+	 * Caller is responsible for allocating sufficient memory for `message`, based
```
This comment is no longer relevant since `message` is input-only here.

christianpaquin replied:
Fixed.

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
+	 * @return OQS_SUCCESS or OQS_ERROR
 	 */
-	OQS_STATUS(*verify)(const OQS_SIG *s, const uint8_t *pub, const uint8_t *msg, const size_t msg_len, const uint8_t *sig, const size_t sig_len);
+	OQS_STATUS (*verify)(uint8_t *message, size_t message_len, const uint8_t *signature, size_t signature_len, const uint8_t *public_key);
```
`message`, `message_len`, and `signature_len` should be marked as `const`.

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
 
 /**
- * Instantiate a new signature object.
+ * Consturcts an OQS_SIG object for a particular algorithm.
```
Typo: constructs

christianpaquin replied:
Fixed; also fixed on nist-branch (PR #393)

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
- * @param algid              The id of the signature algorithm to be instantiated.
- * @return                   A new signature object on success, or NULL on failure.
+ * @param[in] method_name Name of the desired algorithm; one of the names in `OQS_SIG_algs`.
+ * @return An OQS_SIG for the particular algorithm, or `NULL` if the algorithm has been disabled at compile-time.
```
... if the algorithm has been disabled at compile-time or is not recognized.

christianpaquin replied:
We should update kem.h as well...

dstebila replied:
Will do.

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
- * @return          OQS_SUCCESS on success, or OQS_ERROR on failure
+ * Signature generation algorithm.
+ *
+ * Caller is responsible for allocating sufficient memory for `signed_message`,
```
Same comment as above.

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
+	 * @return OQS_SUCCESS or OQS_ERROR
 	 */
-	OQS_STATUS(*sign)(const OQS_SIG *s, const uint8_t *priv, const uint8_t *msg, const size_t msg_len, uint8_t *sig, size_t *sig_len);
+	OQS_STATUS (*sign)(uint8_t *signature, size_t *signature_len, const uint8_t *message, size_t message_len, const uint8_t *secret_key);
```
`message_len` should be marked as const

christianpaquin replied:
Yes; leftovers from copy-pasting from nist-branch

dstebila replied:
Well we do it for uint8_t which can't be changed by the function either, so why one and not the other?

christianpaquin replied:
we do it for uint8_t _pointers_, the const being a promise to the caller that the function won't modify the content. I don't see why that would be useful for a size parameter.

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
+ * @return OQS_SUCCESS or OQS_ERROR
  */
-OQS_API OQS_STATUS OQS_SIG_sign(const OQS_SIG *s, const uint8_t *priv, const uint8_t *msg, const size_t msg_len, uint8_t *sig, size_t *sig_len);
+OQS_STATUS OQS_SIG_sign(const OQS_SIG *sig, uint8_t *signature, size_t *signature_len, const uint8_t *message, size_t message_len, const uint8_t *secret_key);
```
`message_len` should be marked as const

christianpaquin replied:
Yes; leftovers from copy-pasting from nist-branch

dstebila replied:
Well we do it for uint8\_t which can't be changed by the function either, so why one and not the other?

christianpaquin replied:
we do it for uint8_t _pointers_, the const being a promise to the caller that the function won't modify the content. I don't see why that would be useful for a size parameter.

dstebila commented on src/sig/sig.h:
```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
+ * @return OQS_SUCCESS or OQS_ERROR
  */
-OQS_API OQS_STATUS OQS_SIG_verify(const OQS_SIG *s, const uint8_t *pub, const uint8_t *msg, const size_t msg_len, const uint8_t *sig, const size_t sig_len);
+OQS_STATUS OQS_SIG_verify(const OQS_SIG *sig, uint8_t *message, size_t message_len, const uint8_t *signature, size_t signature_len, const uint8_t *public_key);
```
`message, `message_len`, and `signature_len` should be marked as `const`

christianpaquin replied:
Yes; leftovers from copy-pasting from nist-branch

dstebila replied:
Well we do it for uint8_t which can't be changed by the function either, so why one and not the other?

dstebila commented on tests/Makefile.am:
```diff
@@ -1,18 +1,13 @@
 check_PROGRAMS = example_kem speed_kem test_kem \
                       minimal_kex_oqs test_kex \
-                      test_sig \
+                      minimal_sig_oqs test_sig \
```
I'd prefer to use the name `example_sig` instead of `minimal_sig_oqs`, to mirror what we've done for KEMs and as on nist-branch.

christianpaquin replied:
Ah yes, I was wondering if there would be a minimal_kem_oqs to replace the kex one. I can rename that one.

christianpaquin replied:
I migrated the example_sig from the nist-branch.

dstebila commented on tests/test_kem.c:
```diff
@@ -1,4 +1,3 @@
-#include <stdbool.h>
```
Out of curiosity, why did you delete this?

christianpaquin replied:
Wasn't needed in test_sig.c, so after comparing with test_kem.c, I figured it wasn't needed there either.

dstebila commented:
I noticed that the Travis build was failing so I worked through fixing it, but it turned out there were a multitude of flaws in the Travis scripts that meant a bunch of things weren't working properly, so the solution got bigger than I would have liked.  Sorry to pollute your PR with all that extra stuff.

christianpaquin commented:
Let me know when you are done editing the PR, and I'll take back the baton. I also need to add speed_sig to master.

dstebila commented:
Seems to be working now, so baton is yours.

christianpaquin commented:
I think I've addressed all the comments on this PR, with the exception of the "const size_t" comment which I suggest not to do. @dstebila; you can take another look, or provide more comments if I missed something.
